### PR TITLE
docs(api): more docs

### LINKED ISSUE #176: Document token cache
Documentation should go to an extra file, align with other 2.0.0 documentation issues.

Markdown and doctype comments should be in sync.

### PR DESCRIPTION
**Pull request by ISO50**
close #176

close #<ticket number> (if already added automatically by github, delete this line)

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -1,12 +1,120 @@
...
+
+* `tokenName: string` - key of the token as configured in `tokenConfig`
+
+#### Returns
+
+`Promise<Token>` that resolves with a token with configured scopes. In case of error rejects with an error message.
+
+### refreshAllTokens
+
+Triggers the request of a new token. Invalidates the old the cache entry.
```
Comment (Retro64) [inline API.md]: all or one?

File: README.md [hunk 2]
```diff
@@ -17,25 +17,26 @@ Currently the following flows are supported:
 * [Client Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.4)
 * [Resource Owner Password Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.3)
 * [Refresh Token Grant](https://tools.ietf.org/html/rfc6749#section-6)
-* Express middlewares to simplify authentication/authorization
-* `TokenCache` service to manage access tokens in your application
-* Mock tooling for OAuth2.0 endpoints to enable decent unit and integration tests
 
-See [STUPS documentation](http://stups.readthedocs.org/en/latest/user-guide/access-control.html#implementing-a-client-asking-resource-owners-for-permission) and [OAuth2 documentation](https://tools.ietf.org/html/rfc6749) for more information.
+Currently the [Authmosphere JavaScript API](./API.md) supports:
```
Comment (Retro64) [inline README.md]: currently,

File: API.md [hunk 3]
```diff
@@ -1,12 +1,120 @@
...
+
+#### Signature
+
+```ts
+constructor(tokenConfig, oauthConfig: TokenCacheOAuthConfig, options)
+```
+
+#### Arguments
+
+* `tokenConfig: [key: string]: string[]` - the token cache is able to handle multiple tokens. The string array should contain the scopes attached to the token.
```
Comment (bzums) [inline API.md]: Somehow not precise, maybe:

> mapping between a name (representing a token) and the scopes requested for the corresponding token

File: README.md [hunk 4]
```diff
@@ -17,25 +17,26 @@ Currently the following flows are supported:
...
+* [Express middlewares](./API.md#express-tooling) to simplify authentication/authorization
+* [`TokenCache`](./API.md#token-cache) service to manage access tokens in your application
+* OAuth tooling
+  * [`getAccessToken`](./API.md#getaccesstoken) - helper to request access tokens
+  * [`getTokenInfo`](./API.md#gettokeninfo) - help to validate access tokens
+* [Mock tooling](./API.md#mock-tooling) for OAuth2.0 endpoints to enable decent unit and integration tests
 
-* See the [changelog](./CHANGELOG.md) for more information.
-* See the [migration guide](./MIGRATION_GUIDE.md) for more information.
+See [STUPS documentation](http://stups.readthedocs.org/en/latest/user-guide/access-control.html#implementing-a-client-asking-resource-owners-for-permission) and [OAuth2 documentation](https://tools.ietf.org/html/rfc6749) for more information.
```
Comment (Retro64) [inline README.md]: ...about?

File: API.md [hunk 5]
```diff
@@ -1,12 +1,120 @@
...
+* `oauthConfig: TokenCacheOAuthConfig` -
+  Either [`ClientCredentialsGrantConfig`](#clientcredentialsgrantconfig) or [`PasswordCredentialsGrantConfig`](#passwordcredentialsgrantconfig)` plus the additional `tokenInfoEndpoint: string` property that specifies the URL of the token validation endpoint.
```
Comment (bzums) [inline API.md]: wrong quoting (see rendered markdown)

File: API.md [hunk 6]
```diff
@@ -1,12 +1,120 @@
...
 
+## Token Cache
+
+Class to request and cache tokens on client-side.
+
+#### Usage
+
+```typescript
+
+// TokenCacheOptions
```
Comment (bzums) [inline API.md]: remove comment?

File: API.md [hunk 7]
```diff
@@ -1,12 +1,120 @@
...
+  accessTokenEndpoint: 'https://....',
+  credentialsDir: './credentialsDir'
+};
+
+const tokenConfig = {
+  'service-foo': ['foo.read', 'foo.write'],
+  'service-bar': ['bar.read']
+};
+
+// create a new cache
```
Comment (bzums) [inline API.md]: Comment necessary? Should be clear or?

File: API.md [hunk 8]
```diff
@@ -1,12 +1,120 @@
...
+## Token Cache
+
+Class to request and cache tokens on client-side.
+
+#### Usage
+
+```typescript
+
+// TokenCacheOptions
+const cacheOptions = {
```
Comment (bzums) [inline API.md]: Would calling this var `options` make things a little bit less confusing? ^^

File: API.md [hunk 9]
```diff
@@ -1,12 +1,120 @@
...
+    console.log(token.access_token);
+  });
+```
+
+### constructor
+
+#### Signature
+
+```ts
+constructor(tokenConfig, oauthConfig: TokenCacheOAuthConfig, options)
```
Comment (bzums) [inline API.md]: remove type `TokenCacheOAuthConfig`

File: API.md [hunk 10]
```diff
@@ -1,12 +1,120 @@
...
+};
+
+// create a new cache
+const tokenCache = new TokenCache(tokenConfig, oAuthConfig, cacheOptions);
+
+// Get a token from cache
+tokenCache
+  .get('service-foo') // Needs to match with a key from 'tokenConfig'
+  .then((token: Token) => {
+    console.log(token.access_token);
```
Comment (bzums) [inline API.md]: Maybe we should not do the one thing you definitely should not do with a token in our example ;)
In other examples we used a comment like:

`// ...use the token...`

File: API.md (hunk 3 above)
Comment (bzums) [inline API.md]: Sometimes we start the section after `-` with uppercase, sometimes with lowercase, lets do it consistent.

File: API.md [hunk 11]
```diff
@@ -1,12 +1,120 @@
 # Authmosphere API Documentation
 
+## Token Cache
+
+Class to request and cache tokens on client-side.
+
+#### Usage
+
+```typescript
```
Comment (bzums) [inline API.md]: In other examples we added `import` statement (though not everywhere). Shall we be consistent?

File: API.md [hunk 12]
```diff
@@ -1,12 +1,120 @@
...
+* `oauthConfig: TokenCacheOAuthConfig` -
+  Either [`ClientCredentialsGrantConfig`](#clientcredentialsgrantconfig) or [`PasswordCredentialsGrantConfig`](#passwordcredentialsgrantconfig)` plus the additional `tokenInfoEndpoint: string` property that specifies the URL of the token validation endpoint.
+* [`options?: TokenCacheOptions`](./src/types/TokenCacheConfig.ts)
+  * [`cacheConfig?: CacheConfig`](./src/types/TokenCacheConfig.ts)
+    * `percentageLeft: number` - To determine when a token is expired locally (means when to issue a new token): if the token exists for `((1 - percentageLeft) * lifetime)` then issue a new one.
+  * [`logger?: Logger`](#logging)
+
+### get
+
+Returns cached token or requests a new one if lifetime as configured in `cacheOptions.cacheConfig` is expired.
```
Comment (bzums) [inline API.md]: "Returns cached token or requests a new one if lifetime (as configured in `cacheOptions.cacheConfig`) is expired." (brackets)

File: README.md (hunk 2 above)
Comment (ISO50) [reply README.md]: what do you mean?

File: README.md [hunk 13]
```diff
@@ -17,25 +17,26 @@ Currently the following flows are supported:
...
 
-See [STUPS documentation](http://stups.readthedocs.org/en/latest/user-guide/access-control.html#implementing-a-client-asking-resource-owners-for-permission) and [OAuth2 documentation](https://tools.ietf.org/html/rfc6749) for more information.
+Currently the [Authmosphere JavaScript API](./API.md) supports:
 
-## Changelog and Migration
+* [Express middlewares](./API.md#express-tooling) to simplify authentication/authorization
+* [`TokenCache`](./API.md#token-cache) service to manage access tokens in your application
+* OAuth tooling
+  * [`getAccessToken`](./API.md#getaccesstoken) - helper to request access tokens
+  * [`getTokenInfo`](./API.md#gettokeninfo) - help to validate access tokens
```
Comment (bzums) [inline README.md]: "help_er_ to validate..."

File: README.md [hunk 14]
```diff
@@ -46,140 +47,12 @@ import {
...
-```
-
-#### AUTHORIZATION_CODE_GRANT
-
-String constant specifying the Authorization Code Grant type.
-
-#### PASSWORD_CREDENTIALS_GRANT
-
-String constant specifying the Resource Owner Password Credentials Grant type.
+[API](./API.md)
```
Comment (bzums) [inline README.md]: Missing sentence ;)

File: README.md (hunk 4 above)
Comment (ISO50) [reply README.md]: Not exactly sure about 'about'?

File: API.md (hunk 6 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md (hunk 8 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md (hunk 7 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md (hunk 10 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md (hunk 9 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md (hunk 5 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md (hunk 12 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md [hunk 15]
```diff
@@ -1,12 +1,120 @@
...
+`Promise<Token>` that resolves with a token with configured scopes. In case of error rejects with an error message.
+
+### refreshAllTokens
+
+Triggers the request of a new token. Invalidates the old the cache entry.
+
+#### Signature
+
+```ts
+refreshAllTokens(): Promise<TokenMap>
```
[the remark this answers is not in the document]
Comment (ISO50) [reply API.md]: ✅

File: README.md (hunk 13 above)
Comment (ISO50) [reply README.md]: ✅

File: API.md (hunk 3 above)
Comment (ISO50) [reply API.md]: ✅

File: API.md (hunk 1 above)
Comment (ISO50) [reply API.md]: ✅

File: README.md (hunk 2 above)
Comment (ISO50) [reply README.md]: ✅

**Comment by bzums** [comment]
👍

**Comment by ISO50** [comment]
👍

**Pull request merged**